### PR TITLE
vivejest - Vicki, Vera, Jessica and Stephanie (momasy)

### PR DESCRIPTION
**Pull request by vwhwang**
# Assignment Submission: bEtsy

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions. These should be answered by all members of your team, not by a single teammate.

## Reflection

Prompt | Response
--- | ---
Each team member: what is one thing you were primarily responsible for that you're proud of? | Vicki - category many to many relationships and getting comfortable with deploying heroku (ish). Jessica - When one person checks out and another person still has that item in their cart, it updates the inventory to what's available if they try to check out. Vera - Create order items and Cart when the people cannot add more items if there is not enough inventory. Stephanie - side bar only show up on certain pages and carousel. 
Each team member: what is one thing you were primarily responsible for that you would like targeted feedback on? | Vicki - order controller tests. Stephanie - category model test. Jessica - merchant authorization for certain actions . Vera - Order items controller and tests.
How did your team break up the work to be done? | We used trello and the users story to the to-do list. We break it up in the stand up each day. 
How did your team utilize git to collaborate? | We created branches to work on individually and make PR for review before merge.
What did your group do to try to keep your code DRY while many people collaborated on it? | We made some partials and try to reuse style classes. We also use helper methods. 
What was a technical challenge that you faced as a group? | Github / Heroku authentication. 
What was a team/personal challenge that you faced as a group? | carousel and filter status. As a group was hard to make sure we don't step on each other's toes/code. 
What was your application's ERD? (upload this to Google Drive, change the share settings to viewable by everyone with a link, and include a link) | see trello 
What is your Trello URL? | https://trello.com/b/Jn7dIbHr/vivejest
What is the Heroku URL of your deployed application? | https://momasy.herokuapp.com/

### CODE REVIEW

File: app/assets/stylesheets/application.scss [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+.btn-success:hover{
+  background-color:  olive;
+  border-color: olive;
+}
+
+.thead-table, .tbody{
+  color: #444444;
+}
+
+@media screen and (max-width: 600px) {
```
Comment (kaidamasaki) [inline app/assets/stylesheets/application.scss:256]: Oooh, responsive design!

File: app/controllers/order_items_controller.rb [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+      # params[:quantity].to_i to add up the quantity
+      if (current_item[:quantity] + params[:quantity].to_i) > @product.inventory
+        flash[:error] = "You cannot add more items than are in stock."
+        redirect_to product_path(@product.id)
+      else
+        current_item[:quantity] += params[:quantity].to_i
+        current_item.save
+        flash[:success] = "Successfully updated order item"
+        redirect_to order_path(@order_id)
+      end
```
Comment (kaidamasaki) [inline app/controllers/order_items_controller.rb:32]: This logic should probably live in the model.

File: app/controllers/order_items_controller.rb [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    })
+
+    if order_item.save
+      flash[:success] = "Item added to order"
+      redirect_to order_path(@order_id)
+    else
+      flash[:error] = "Could not add item to order"
+      redirect_to order_path(@order_id)
+    end
+  end
```
Comment (kaidamasaki) [inline app/controllers/order_items_controller.rb:113]: This should be a factory method on `OrderItem`.

File: app/controllers/orders_controller.rb [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        end
+        inventory_errors += 1
+      end
+    end
+
+    if inventory_errors > 0
+      flash[:error] = "Some of the items in your cart are no longer in stock. We have updated your cart to reflect the current quantity."
+      redirect_to order_path(params[:id])
+      return
+    end
```
Comment (kaidamasaki) [inline app/controllers/orders_controller.rb:68]: This logic should have been in a method on `Order`.

**Comment by kaidamasaki** [comment]
# bEtsy



## Functional Requirements: Manual Testing

Workflow | yes / no
---   | ---
Deployed to Heroku | ✔️
**Before logging in** |
Browse all products, by category, by merchant | ✔️
Leave a review | ✔️
Verify unable to create a new product | ✔️
**After logging in** | |
Create a category | ✔️
Create a product in that category with stock 10 | ✔️
Add the product you created to your cart | ✔️
Add it again (should update quantity) | ✔️
Verify unable to increase quantity beyond stock | ✔️
Add another merchant's product | ✔️
Check out | Broken on deployed copy.
Check that stock was reduced | Checkout did not work.
Change order-item's status on dashboard | ✔️
Verify unable to leave a review for your own product | ✔️
Verify unable to edit another merchant's product by manually editing URL | ✔️
Verify unable to see another merchant's dashboard by manually editing URL | ✔️

## Major Learning Goals/Code Review

Criteria | yes / no
---  | ---
90% reported coverage for all controller and model classes using SimpleCov | ✔️ 96.66%!
**Routes** |
No un-needed routes generated (check `reviews`) | ✔️
Routes not overly-nested (check products and merchants) | ✔️
Merchant dashboard and cart page use a non-parameterized routes (should pull merchant or cart ID from session) | ✔️
**Controllers** |
Controller-filter to require login by default | ✔️
Helper methods or filters to find logged-in user, cart, product, etc | ✔️
No excessive business logic | A moderate amount of business logic lives in the controller.
**Business logic that ought to live in the model** |
Add / remove / update product on order | ✔️
Checkout -> decrease inventory | ✔️
Merchant's total revenue | ✔️
Find all orders for this merchant (instance method on `Merchant`) | ✔️
**Selected Model Tests** |
Add item to cart:<br>  - Can add a good product<br>  - Can't add a product w/o enough stock<br>  - Can't add a retired product<br>  - Can't add to an order that's not in cart mode<br>  - Logic specific to this implementation | (Only the specific logic.)
Get orders for this merchant:<br>  - Includes all orders from this merchant<br>  - Doesn't include orders from another merchant<br>  - Orders are not included more than once<br>  - Does something reasonable when there are no orders for this merchant | ✔️ (Mostly.)
**Selected Controller Tests** |
Add item to cart:<br>  - Empty cart (should be created)<br>  - Cart already exists (should add to same order)<br>  - Product already in cart (should update quantity)<br>  - Bad product ID, product is retired, quantity too high, or something like that (error) | ✔️ (Mostly.)
Leave a review:<br>  - Works when not logged in<br>  - Works when logged in as someone other than the product's merchant<br>  - Doesn't work if logged in as this product's merchant<br>  - Doesn't work if validations fail | ✔️ (Mostly.)



## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅

## Overall Feedback

Great work overall! You've built a fully functional web store from top to bottom. This represents a huge amount of work, and you should be proud of yourselves!.

I am particularly impressed by the way that you styled your site and your (super high!) test coverage.

I do see some room for improvement around manual testing of edge cases and moving functionality into models.

bEtsy is a huge project on a very short timeline, and this feedback should not at all diminish the magnitude of what you've accomplished. Keep up the hard work!

Overall Feedback: Meets or Exceeds Standard



## Only the person who submitted the PR will get an email about this feedback. Please let the rest of your team know about it.
